common/config.go

```diff
@@ -47,81 +46,52 @@ func ChooseString(vals ...string) string {
 // a completely valid URL. For example, the original URL might be "local/file.iso"
 // which isn't a valid URL. DownloadableURL will return "file:///local/file.iso"
 func DownloadableURL(original string) (string, error) {
-	if runtime.GOOS == "windows" {
-		// If the distance to the first ":" is just one character, assume
-		// we're dealing with a drive letter and thus a file path.
-		idx := strings.Index(original, ":")
-		if idx == 1 {
-			original = "file:///" + original
-		}
-	}
 
-	url, err := url.Parse(original)
-	if err != nil {
-		return "", err
+	// Verify that the scheme is something we support in our common downloader.
```

Comment:
I re-added the FTP test above (since we have removed FTP support from this test) and it causes a test failure. Moving the validation up here and preventing it from failing out if the validation fails makes this clause _much_ more permissive.  If I were to make a scheme of "asdf://host.com/path" it would pass this validation without throwing an error.

Reply:
Sure whichever you want to do.

When I first made this patch, the way that I isolated the logic between config.go's DownloadableURL and download.go's protocols was by making DownloadableURL only do what the documentation had described. Namely, to take the input and force it into an rfc-compliant URI. I left the original checks for the scheme because it was already there originally even though the documentation didn't mention that the job of the function was to validate the uri for the downloader.

`<IMO>`the usage of DownloadableURL to validate whether a protocol is valid or not is against what the documentation describes and the checking of the protocols in that function is a hack because it can't possibly have any idea what protocols are actually implemented by the downloader. (That's why I left a /* XXX */ reference in downloader to ensure people don't forget to specify the protocols in both places.) So some of the tests that assume DownloadableURL should decide that a bad scheme/protocol is the same as a bad url is a bad assumption.`</IMO>`

At the very least, we should update the documentation for DownloadableURL to point out that it's job isn't just to convert a path to an rfc-compliant `file://` uri, but also to validate that the specified protocol is valid as well.

Reply:
I see where you're coming from there.  One of the problems with an old, organically-grown codebase is that there's often drift in intended function that somehow doesn't make it into the docstrings.  At this point, I think there are a couple of places in the code that expect DownloadableURL to do some basic validation; I would be happy with either a) making sure that the protocol is being validated as it was in DownloadableURL's previous life and updating the docstring or b) moving said protocol validation into a separate ValidateProtocol function inside the common library, and then bundling DownloadableURL and ValidateProtocol together into some DownloadableAndValidatedURL function that we use in all the places we currently call DownloadableURL.